scheds/rust/scx_layered/src/bpf/main.bpf.c

```diff
@@ -157,6 +159,126 @@ struct {
 	__uint(max_entries, 1);
 } cpu_ctxs SEC(".maps");
 
+struct layer_cpumask_wrapper {
+	struct bpf_cpumask __kptr *cpumask;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__type(key, u32);
+	__type(value, struct layer_cpumask_wrapper);
+	__uint(max_entries, 2);
+} percpu_tmpmask SEC(".maps");
+
+static s32 create_save_cpumask(struct bpf_cpumask **kptr)
+{
+	struct bpf_cpumask *cpumask;
+
+	cpumask = bpf_cpumask_create();
+	if (!cpumask) {
+		scx_bpf_error("Failed to create cpumask");
+		return -ENOMEM;
+	}
+
+	cpumask = bpf_kptr_xchg(kptr, cpumask);
+	if (cpumask) {
+		scx_bpf_error("kptr already had cpumask");
+		bpf_cpumask_release(cpumask);
+	}
+
+	return 0;
+}
+
+
+static inline struct bpf_cpumask *lookup_percpu_tmpmask(int ind)
+{
+	struct layer_cpumask_wrapper *wrapper; 
+
+	wrapper = bpf_map_lookup_elem(&percpu_tmpmask, &ind);
+	if (!wrapper)
+		return NULL;
+
+	return wrapper->cpumask;
+}
+
+static inline int init_percpu_tmpmask_one(s32 cpu, int ind)
+{
+	struct layer_cpumask_wrapper *wrapper;
+
+	wrapper = bpf_map_lookup_percpu_elem(&percpu_tmpmask, &ind, cpu);
+	if (!wrapper)
+		return -EINVAL;
+
+	return create_save_cpumask(&wrapper->cpumask);
+}
+
+static inline int init_percpu_tmpmasks(s32 cpu)
+{
+	int ret;
+	int i;
+
+	bpf_for(i, 0, 2) {
+		ret = init_percpu_tmpmask_one(cpu, i);
+		if (ret)
+			return ret;
+	}
+
+	return 0;
+}
+
+static __always_inline 
+struct bpf_cpumask *idle_cpumask_unprotected(struct cpu_ctx *cpuc, const struct cpumask *layermask)
+{
+	const struct cpumask __kptr *idlemask;
+	struct bpf_cpumask __kptr *cpumask;
+
+	if (!(cpumask = lookup_percpu_tmpmask(0))) {
+		scx_bpf_error("cpu context has no mask");
+		return NULL;
+	}
+
+	idlemask = scx_bpf_get_idle_cpumask();
+	if (!idlemask || !unprotected_cpumask) {
+		scx_bpf_put_idle_cpumask(idlemask);
+		return NULL;
+	}
+
+	/* Candidate CPUs are either in this layer or not in a protected layer. */
+	bpf_cpumask_or(cpumask, layermask, cast_mask(unprotected_cpumask));
+
+	bpf_cpumask_and(cpumask, cast_mask(cpumask), idlemask);
```

Comment:
The `unprotected` mask only changes when layer CPU allocations are updated and is only needed when a layer can't find an idle CPU that it owns. Maybe I'm confused but I don't think we need to be computing these masks dynamically. We'll only overflow into unprotected CPUs when there are no owned CPUs, so we don't necessarily need to or them with owned CPUs and even if we did we can calculate the result and cache it using `cached_cpus` and friends.